chef/cookbooks/utils/libraries/role_recipe.rb

```diff
@@ -14,9 +14,44 @@
 #
 
 module CrowbarRoleRecipe
+  # Blacklist mechanism which uses role recipes to allow skipping of
+  # recipes in certain situations.
+  def self.skip_role(node, barclamp, role)
+    # If we're applying the nova-ha-compute batch, skip other roles to avoid
+    # chef-client drift between nodes (since sync marks are only reset before
+    # the first batch) - https://bugzilla.suse.com/show_bug.cgi?id=1004758
+    if node["crowbar"]["applying_for"].key?("nova") &&
+        node["crowbar"]["applying_for"]["nova"].include?("nova-ha-compute") &&
+        role != "nova-ha-compute"
+      return "not required in nova-ha-compute batch"
+    end
+
+    nil
+  end
+
   # Helper to decide whether a role should be run for a node, depending on the
   # state of the node.
   def self.node_state_valid_for_role?(node, barclamp, role)
+    if node[:state] == "applying"
+      # We always want deployer-client, both for heartbeat but also because it
+      # sets up the ability to use the barclamp library
+      return true if role == "deployer-client"
+
+      skip_reason = skip_role(node, barclamp, role)
+      if skip_reason
+        roles = node["crowbar"]["applying_for"].collect { |k, v| v }.flatten.sort.uniq
+        if roles.include? role
+          raise "BUG: skip_role tried to skip #{role} whilst applying " \
+                "batch #{roles} for #{barclamp}"
+        end
+        Chef::Log.info("Skipping role #{role} whilst applying " \
+                       "batch #{roles} for #{barclamp}: #{skip_reason}")
+        return false
+      end
+
+      return true
```

Comment:
I'm still struggling to understand you.

> if you apply the IPMI barclamp, the ipmi-discover role recipe should be skipped

Right, and currently `role_ipmi.rb` has:

    if CrowbarRoleRecipe.node_state_valid_for_role?(node, "ipmi", "ipmi")
      if ["discovering", "hardware-installing", "readying"].include? node[:state]
        include_recipe "ipmi::ipmi-discover"
      end

> Assuming that we generalize the code here,

What do you mean by "generalize the code"?  That's very vague.

> then it will not be the case.

It would, unless you are talking about modifying `role_ipmi.rb`?  Without any modifications to `#node_state_valid_for_role`, this role recipe will *never* run `ipmi-discover` on nodes in the `applying` state.

> The point is that the check that we add here is an additional check, and it should not cause the other checks to be skipped.

Why would it ever cause that?